jax/random.py

```diff
@@ -946,6 +946,25 @@ def _gumbel(key, shape, dtype):
   return -np.log(-np.log(
       uniform(key, shape, dtype, minval=np.finfo(dtype).eps, maxval=1.)))
 
+def categorical(key, logits, axis=-1, shape=()):
+  """Sample random values from categorical distributions.
+
+  Args:
+    key: a PRNGKey used as the random key.
+    logits: Unnormalized log probabilities of the categorical distribution(s) to sample from, so that
+    `softmax(logits, axis)` gives the corresponding probabilities.
+    axis: Specifies the axis along which logits belong to the same categorical distribution.
+    shape: Specifies how many samples to take per categorical distribution and in what shape to arrange them.
+
+  Returns:
+    Samples of shape `shape + tuple(onp.delete(logits.shape, axis))`.
+    The first len(shape) dimensions specify which sample,
+    while the last len(logits.shape) - 1 dimensions make up the shape per sample.
+    """
+  if axis >= 0:
+    axis += len(shape)
+
+  return np.argmax(gumbel(key, shape + logits.shape, logits.dtype) + logits, axis=axis)
```

Comment:
I thought that we would need to broadcast `logits` here: something like having `shape=None` (or `shape=()`) by default
```
if shape is None:
    shape = logits.shape[:-1]
logits = np.broadcast_to(logits, shape)
return np.argmax(gumbel(key, logits.shape, logits.dtype) + logits, axis=-1)
```

Reply:
@fehiepsi Could you give an example where the current implementation misbehaves? The default for `shape` is already `()`. The output shape is the concatenation of `shape` and `logits.shape[:-1]` for `axis=-1`.

Reply:
It'll work as intended. I just meant to have `shape` to be output shape like other samplers. (though I don't have a strong opinion on this).

Reply:
@fehiepsi Now I see what you mean, i. e. `dirichlet` in the same file is doing this already. Why would we require the user to repeat the shape of `alpha.shape[:-1]`? I. e.

```
dirichlet(PRNGKey(0), np.array([[1., 1.], [1., 1.]]), shape=(2,)).shape
```
is currently `(2, 2)`, but I think it might be better to have it `(2, 2, 2)`, so that `shape` represents the shape of samples we draw from _every_ specified distribution. This would be consistent with TF probability's [`sample`](https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/Distribution#sample) that also only requires this shape (they call it `sample_shape`).

Reply:
If I remember correctly, then `shape = sample_shape + batch_shape` to match numpy.random (or scipy.stats) behavior. I agree that it is not necessarily required and can be simplified with `shape = sample_shape`. But matching numpy behavior is convenient for users IMO. Maybe @mattjj will have more opinions about this. :)

Reply:
I have now forgotten everything, and I unfortunately don't have time to page this back in right this minute. But I think our motivation was summarized in #1515.

Reply:
My thinking is that we should keep things in the file mutually consistent. So please make the `shape` semantics match the other samplers for vector-range random variables in this file. If we got that `shape` semantics wrong and want to revise it, we should have that discussion and fix in a separate PR so as not to block this one getting in.

WDYT? Does that make sense?

Reply:
I think #1515 discussed `shape = batch_shape + event_shape` (where event_shape is 0D for scalar-range RV or 1D for vector-range RV) or `shape = batch_shape`. This categorical distribution will have scalar-range so we might just skip discussing vector-range RVs such as Dirichlet or MVN for now. The `shape` semantics of the categorical distribution in this PR is different from other scalar-range RVs such as Bernoulli. Here, we have another notion `sample_shape`, which is introduced in tensorflow/pytorch folks. It can be convenient but might not be necessarily required. If we would like to change the semantics here, then we will need to change the behavior of other samplers (`bernoulli`, `gamma`, `beta`, `pareto`, `randint`, `t`,...) too.

So I propose to make the behavior of categorical sampler match other scalar-range samplers: `shape` is the output shape. It will be helpful for numpy users that they don't have to learn the new shape semantics.  For those who want the additional tensorflow behavior of `sample_shape`, they just need to broadcast `logits` or `shape` to the required value. It is not a big deal IMO.

Reply:
My mistake: I was thinking of the one-hot version of categorical, but this is the scalar one. Good point!

I agree with everything @fehiepsi wrote.

Reply:
@mattjj @fehiepsi Now updated behavior to `shape = sample_shape + batch_shape` for consistency / to allow merge. However, even after reading all of https://github.com/google/jax/pull/1515, I am still very much against this convention. It

- requires the user to specify duplicate information,
- makes the API harder to understand IMO),
- makes the implementation more complex,
- deviates from the (widely used) TF/pytorch convention.

Therefore, I'd strongly prefer `shape = sample_shape` for all of `categorical`, `bernoulli`, `gamma`, `beta`, `pareto`, `randint`, `t`, etc. 

I'd volunteer to update all of them.

Reply:
Thanks for the change!

We should have the convention discussion for all samplers in a new issue, because I'm not sure I agree yet, though I can be convinced. In short, redundant information isn't always a bad thing, especially if it means being explicit rather than implicit since labeling only partial information means other information is implicit; the API being harder to understand is subjective without referring to specific principles or criteria, like explicit vs implicit; the implementation complexity difference is pretty small, like a couple lines, and should be measured additively not proportionally; and not all TF/PyTorch conventions are good ones, though they can set our priors and perhaps point us to others we should ask (like the TFP devs). IMO the strongest argument for a change in the convention will be from first principles.

I am willing to be convinced though, and I'm especially eager to get @fehiepsi 's and @neerajprad 's input!

Reply:
Sorry, I'm coming to this late. My thought is that the current design of having `shape` denote the final sample shape is probably going to cause the least surprise to existing numpy/scipy users. It is also easy enough to build light wrappers around these samplers like we have done in NumPyro ([example](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/distributions/continuous.py)) for a more opinionated distributions API. We have followed PyTorch's convention so as to be compatible for Pyro users, which itself closely follows the TF distributions design. Unless we would like to build a distributions module like in TFP, `torch.distributions` or `scipy.stats`, I think any consistent convention that JAX follows should be fine.

I can expand on why we decided to go with the TF distributions API, specially with regard to distinguishing between the different distribution dimensions. Note that scipy doesn't make this distinction consistently, and we noticed many small inconsistencies and pain points when we were following their API in Pyro in the very early days. In TF and PyTorch, when a distribution is created, it has a `batch_shape` and an `event_shape`, the number of event dims being fixed for a distribution class. e.g. Dirichlet has a single event dim and univariate distributions have none. Batch dims correspond to the independent (not necessarily identical) draws from the distribution, and event dims correspond to a single draw (where elements in the event dim are possibly dependent). Additionally, we can specify a `sample_shape` to the distribution's `sample` method for iid draws from the distribution. It makes sense to place these from left to right in order of whether they are likely to change, i.e. `sample_shape + batch_shape + event_shape`, where the `event_shape` (also, `batch_shape`) are fixed on construction. One reason for such a placement is that broadcasting of tensors happens on the left, i.e. shapes `(2,), (3, 2)` are broadcastable, but not `(2,), (2, 3)`, so this makes it easy to just expand the distribution's parameters by tacking on sample dims on the left to draw iid samples. The tensor returned by distribution's `log_prob` has the shape `sample_shape + batch_shape` (the event dims are summed out; which is another reason to locate event dims on the right and not in between the sample and batch dims).

Given this design, it made sense for distribution's `sample` method to only specify the iid dimensions on the left for the reasons that @JuliusKunze mentioned above. This has worked well for us in PyTorch, but I recognize that there may be other design possibilities and this requires the user to keep this distribution shapes model in mind, as @fehiepsi said. The API discussion may become more relevant if we are looking to have a separate distributions module like `scipy.stats` or an external distributions library like TFP.